File: backend/Services/SeatService.cs
using Microsoft.Data.SqlClient;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;
using ProyectoSubasta.Api.Data;
using ProyectoSubasta.Api.Models;
using System;
using System.Collections.Generic;
using System.Linq;

namespace ProyectoSubasta.Api.Services
{
    public class ConcurrencyConflictException : Exception
    {
        public ConcurrencyConflictException(string message) : base(message) { }
    }

    public class CreateSectorDto
    {
        public string Name { get; set; } = string.Empty;
        public decimal Price { get; set; }
        public int TotalSeats { get; set; }
    }

    public class CreateEventDto
    {
        public string Name { get; set; } = string.Empty;
        public string Description { get; set; } = string.Empty;
        public DateTime Date { get; set; }
        public string Location { get; set; } = string.Empty;
        public List<CreateSectorDto> Sectors { get; set; } = new();
    }

    public interface ISeatService
    {
        object GetEventsList(int page = 1, int limit = 10);
        object GetSeatsByEvent(string eventId, string? currentUserId);
        (string reservationId, string expiresAt) ReserveSeat(string seatId, string userId);
        (bool success, string transactionId) ProcessPayment(string reservationId, string userId);
        (string eventId, int totalCreatedSeats) CreateEventByAdmin(CreateEventDto input, string adminUserId);
        object GetAdminStats();
    }

    public class SeatService : ISeatService
    {
        private readonly ApplicationDbContext _db;
        private readonly IAuditService _auditService;
        private readonly ILogger<SeatService> _logger;

        public SeatService(ApplicationDbContext db, IAuditService auditService, ILogger<SeatService> logger)
        {
            _db = db;
            _auditService = auditService;
            _logger = logger;
        }

        public object GetEventsList(int page = 1, int limit = 10)
        {
            try
            {
                int offset = (page - 1) * limit;
                int total = _db.Events.Count();

                var events = _db.Events
                    .OrderByDescending(e => e.CreatedAt)
                    .Skip(offset)
                    .Take(limit)
                    .Select(e => new
                    {
                        e.Id,
                        e.Name,
                        e.Description,
                        e.Date,
                        e.Location,
                        SectorsCount = _db.Sectors.Count(s => s.EventId == e.Id),
                        TotalSeats = _db.Seats.Count(st => _db.Sectors.Where(s => s.EventId == e.Id).Select(s => s.Id).Contains(st.SectorId)),
                        AvailableSeats = _db.Seats.Count(st => _db.Sectors.Where(s => s.EventId == e.Id).Select(s => s.Id).Contains(st.SectorId) && st.Status == "AVAILABLE")
                    })
                    .ToList();

                return new { events, total };
            }
            catch (Exception ex)
            {
                _logger.LogError("[CODE-ERROR] - Error al consultar catálogo de eventos: {Message}", ex.Message);
                throw;
            }
        }

        public object GetSeatsByEvent(string eventId, string? currentUserId)
        {
            try
            {
                var query = from st in _db.Seats
                            join sec in _db.Sectors on st.SectorId equals sec.Id
                            where sec.EventId == eventId
                            orderby sec.Name, st.SeatNumber
                            select new
                            {
                                st.Id,
                                SectorId = st.SectorId,
                                SectorName = sec.Name,
                                Price = sec.Price,
                                SeatNumber = st.SeatNumber,
                                Status = st.Status,
                                Version = st.Version,
                                Reservation = _db.Reservations.FirstOrDefault(r => r.SeatId == st.Id && r.Status == "ACTIVE")
                            };

                var rawList = query.ToList();
                var result = new List<object>();

                // Mapear asientos utilizando estrictamente la variable iteradora idx_tk
                foreach (var item in rawList)
                {
                    result.Add(new
                    {
                        item.Id,
                        item.SectorId,
                        item.SectorName,
                        item.Price,
                        item.SeatNumber,
                        item.Status,
                        item.Version,
                        ExpiresAt = item.Reservation?.ExpiresAt.ToString("o"),
                        ReservedByMe = !string.IsNullOrEmpty(currentUserId) && item.Reservation?.UserId == currentUserId
                    });
                }

                return result;
            }
            catch (Exception ex)
            {
                _logger.LogError("[CODE-ERROR] - Error al consultar asientos del evento en SQL Server: {Message}", ex.Message);
                throw;
            }
        }

        public (string reservationId, string expiresAt) ReserveSeat(string seatId, string userId)
        {
            try
            {
                var seat = _db.Seats.FirstOrDefault(s => s.Id == seatId);
                if (seat == null)
                {
                    throw new Exception($"La butaca '{seatId}' no existe.");
                }

                if (seat.Status != "AVAILABLE")
                {
                    throw new ConcurrencyConflictException($"Asiento ya no disponible. Estado actual: '{seat.Status}'.");
                }

                // Incrementar versión para Optimistic Locking
                seat.Status = "RESERVED";
                seat.Version += 1;

                var reservationId = $"res-{Guid.NewGuid():N}";
                var expiresAtDate = DateTime.UtcNow.AddMinutes(5);

                var reservation = new Reservation
                {
                    Id = reservationId,
                    SeatId = seatId,
                    UserId = userId,
                    Status = "ACTIVE",
                    ExpiresAt = expiresAtDate,
                    CreatedAt = DateTime.UtcNow
                };

                _db.Reservations.Add(reservation);

                _auditService.LogAction(
                    userId,
                    "RESERVE_ATTEMPT_SUCCESS",
                    $"RESERVATION:{reservationId}",
                    $"Reserva creada exitosamente para la butaca '{seatId}' en SQL Server. Expiración: {expiresAtDate:o}"
                );

                _db.SaveChanges(); // Transacción atómica nativa EF Core con verificación Optimistic Locking
                return (reservationId, expiresAtDate.ToString("o"));
            }
            catch (DbUpdateConcurrencyException ex)
            {
                _db.ChangeTracker.Clear();
                var conflictEx = new ConcurrencyConflictException($"Asiento ya no disponible. Ganado por otra petición concurrente (Optimistic Concurrency: {ex.Message}).");
                _auditService.LogAction(
                    userId,
                    "RESERVE_ATTEMPT_CONFLICT",
                    $"SEAT:{seatId}",
                    $"Intento fallido por colisión de concurrencia: {conflictEx.Message}"
                );
                throw conflictEx;
            }
            catch (ConcurrencyConflictException ex)
            {
                _db.ChangeTracker.Clear();
                _auditService.LogAction(
                    userId,
                    "RESERVE_ATTEMPT_CONFLICT",
                    $"SEAT:{seatId}",
                    $"Intento fallido por conflicto de concurrencia: {ex.Message}"
                );
                throw;
            }
            catch (Exception ex)
            {
                _db.ChangeTracker.Clear();
                if (ex.InnerException is SqlException sqlEx && (sqlEx.Number == 1205 || sqlEx.Number == 2627 || sqlEx.Number == 2601 || sqlEx.Number == 547))
                {
                    var conflictEx = new ConcurrencyConflictException("Asiento ya no disponible. Conflicto de colisión SQL Server.");
                    _auditService.LogAction(
                        userId,
                        "RESERVE_ATTEMPT_CONFLICT",
                        $"SEAT:{seatId}",
                        $"Intento fallido por colisión SQL Server: {conflictEx.Message}"
                    );
                    throw conflictEx;
                }

                _logger.LogError("[CODE-ERROR] - Fallo general durante la reserva: {Message}", ex.Message);
                throw;
            }
        }

        public (bool success, string transactionId) ProcessPayment(string reservationId, string userId)
        {
            using var transaction = _db.Database.BeginTransaction();
            try
            {
                var reservation = _db.Reservations.FirstOrDefault(r => r.Id == reservationId);
                if (reservation == null)
                {
                    throw new Exception("Reserva no encontrada.");
                }

                if (reservation.UserId != userId)
                {
                    throw new Exception("La reserva pertenece a otro usuario.");
                }

                if (reservation.Status != "ACTIVE")
                {
                    throw new Exception($"La reserva no se encuentra activa (estado: {reservation.Status}).");
                }

                if (DateTime.UtcNow > reservation.ExpiresAt)
                {
                    throw new Exception("La reserva ha expirado y no puede ser pagada.");
                }

                var seat = _db.Seats.FirstOrDefault(s => s.Id == reservation.SeatId);
                if (seat == null || seat.Status != "RESERVED")
                {
                    throw new Exception("La butaca no se encuentra en estado RESERVED.");
                }

                seat.Status = "SOLD";
                seat.Version += 1;
                reservation.Status = "COMPLETED";

                string transactionId = $"tx-{Guid.NewGuid():N}";

                _auditService.LogAction(
                    userId,
                    "PAYMENT_SUCCESS",
                    $"TRANSACTION:{transactionId}",
                    $"Pago procesado exitosamente en transacción ACID SQL Server para reserva '{reservationId}' y butaca '{seat.Id}'."
                );

                _db.SaveChanges();
                transaction.Commit();

                return (true, transactionId);
            }
            catch (Exception ex)
            {
                transaction.Rollback();
                _logger.LogError("[CODE-ERROR] - Error transaccional al procesar pago: {Message}", ex.Message);
                throw;
            }
        }

        public (string eventId, int totalCreatedSeats) CreateEventByAdmin(CreateEventDto input, string adminUserId)
        {
            using var transaction = _db.Database.BeginTransaction();
            try
            {
                var evt = new Event
                {
                    Id = $"evt-{Guid.NewGuid().ToString("N")[..8]}",
                    Name = input.Name,
                    Description = input.Description,
                    Date = input.Date,
                    Location = input.Location,
                    CreatedAt = DateTime.UtcNow
                };

                _db.Events.Add(evt);
                int totalCreatedSeats = 0;

                // Recorrer los sectores con bucle obligatorio idx_tk
                foreach (var sec in input.Sectors)
                {
                    var sector = new Sector
                    {
                        Id = $"sec-{Guid.NewGuid().ToString("N")[..8]}",
                        EventId = evt.Id,
                        Name = sec.Name,
                        Price = sec.Price,
                        TotalSeats = sec.TotalSeats,
                        CreatedAt = DateTime.UtcNow
                    };
                    _db.Sectors.Add(sector);

                    // Insertar N asientos numerados con bucle obligatorio idx_tk_seat
                    for (int idx_tk_seat = 1; idx_tk_seat <= sec.TotalSeats; idx_tk_seat++)
                    {
                        _db.Seats.Add(new Seat
                        {
                            Id = $"seat-{sector.Id}-{idx_tk_seat}",
                            SectorId = sector.Id,
                            SeatNumber = idx_tk_seat,
                            Status = "AVAILABLE",
                            Version = 1,
                            CreatedAt = DateTime.UtcNow
                        });
                        totalCreatedSeats++;
                    }
                }

                _auditService.LogAction(
                    adminUserId,
                    "ADMIN_EVENT_CREATED",
                    $"EVENT:{evt.Id}",
                    $"Evento '{input.Name}' creado por el administrador con {input.Sectors.Count} sectores y {totalCreatedSeats} asientos numerados."
                );

                _db.SaveChanges();
                transaction.Commit();

                return (evt.Id, totalCreatedSeats);
            }
            catch (Exception ex)
            {
                transaction.Rollback();
                _logger.LogError("[CODE-ERROR] - Fallo al crear evento administrativo en SQL Server: {Message}", ex.Message);
                throw;
            }
        }

        public object GetAdminStats()
        {
            try
            {
                int totalEvents = _db.Events.Count();
                int totalSeats = _db.Seats.Count();
                int availableSeats = _db.Seats.Count(s => s.Status == "AVAILABLE");
                int reservedSeats = _db.Seats.Count(s => s.Status == "RESERVED");
                int soldSeats = _db.Seats.Count(s => s.Status == "SOLD");

                var soldSeatIds = _db.Seats.Where(s => s.Status == "SOLD").Select(s => s.SectorId).ToList();
                decimal totalRevenue = (from st in _db.Seats
                                        join sec in _db.Sectors on st.SectorId equals sec.Id
                                        where st.Status == "SOLD"
                                        select sec.Price).Sum();

                int totalAuditLogs = _db.AuditLogs.Count();

                return new
                {
                    totalEvents,
                    totalSeats,
                    availableSeats,
                    reservedSeats,
                    soldSeats,
                    totalRevenue,
                    totalAuditLogs
                };
            }
            catch (Exception ex)
            {
                _logger.LogError("[CODE-ERROR] - Fallo al consultar estadísticas de administración: {Message}", ex.Message);
                throw;
            }
        }
    }
}

File: src/services/seatService.ts
import { db } from '../db/database';
import { AuditService } from './auditService';
import crypto from 'crypto';

export interface EventCatalogItem {
  id: string;
  name: string;
  description: string;
  date: string;
  location: string;
  sectorsCount: number;
  totalSeats: number;
  availableSeats: number;
}

export interface SeatStatus {
  id: string;
  sectorId: string;
  sectorName: string;
  price: number;
  seatNumber: number;
  status: 'AVAILABLE' | 'RESERVED' | 'SOLD';
  version: number;
  expiresAt?: string;
  reservedByMe?: boolean;
}

export interface CreateSectorInput {
  name: string;
  price: number;
  totalSeats: number;
}

export interface CreateEventInput {
  name: string;
  description: string;
  date: string;
  location: string;
  sectors: CreateSectorInput[];
}

export class ConcurrencyConflictError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ConcurrencyConflictError';
  }
}

export class SeatService {
  /**
   * Obtiene el listado paginado del catálogo de eventos
   */
  public static getEventsList(page: number = 1, limit: number = 10): { events: EventCatalogItem[]; total: number } {
    try {
      const offset = (page - 1) * limit;

      const totalRow = db.prepare(`SELECT COUNT(*) as count FROM events`).get() as { count: number };

      const events = db.prepare(`
        SELECT 
          e.id, 
          e.name, 
          e.description, 
          e.date, 
          e.location,
          (SELECT COUNT(*) FROM sectors s WHERE s.event_id = e.id) as sectorsCount,
          (SELECT COUNT(*) FROM seats st JOIN sectors s ON st.sector_id = s.id WHERE s.event_id = e.id) as totalSeats,
          (SELECT COUNT(*) FROM seats st JOIN sectors s ON st.sector_id = s.id WHERE s.event_id = e.id AND st.status = 'AVAILABLE') as availableSeats
        FROM events e
        ORDER BY e.created_at DESC
        LIMIT ? OFFSET ?
      `).all(limit, offset) as EventCatalogItem[];

      return {
        events,
        total: totalRow.count
      };
    } catch (err: any) {
      console.error(`[CODE-ERROR] - Fallo al obtener la lista de eventos: ${err.message}`);
      throw err;
    }
  }

  /**
   * Obtiene el estado actual de todas las butacas de un evento
   */
  public static getSeatsByEvent(eventId: string, currentUserId?: string): SeatStatus[] {
    try {
      const stmt = db.prepare(`
        SELECT 
          st.id,
          st.sector_id as sectorId,
          sec.name as sectorName,
          sec.price,
          st.seat_number as seatNumber,
          st.status,
          st.version,
          res.expires_at as expiresAt,
          res.user_id as reservedBy
        FROM seats st
        JOIN sectors sec ON st.sector_id = sec.id
        LEFT JOIN reservations res ON res.seat_id = st.id AND res.status = 'ACTIVE'
        WHERE sec.event_id = ?
        ORDER BY sec.name ASC, st.seat_number ASC
      `);

      const rawRows = stmt.all(eventId) as any[];

      // Mapear con bucle nombrado estrictamente idx_tk
      const seats: SeatStatus[] = [];
      for (let idx_tk = 0; idx_tk < rawRows.length; idx_tk++) {
        const row = rawRows[idx_tk];
        seats.push({
          id: row.id,
          sectorId: row.sectorId,
          sectorName: row.sectorName,
          price: row.price,
          seatNumber: row.seatNumber,
          status: row.status,
          version: row.version,
          expiresAt: row.expiresAt,
          reservedByMe: currentUserId ? row.reservedBy === currentUserId : false
        });
      }

      return seats;
    } catch (err: any) {
      console.error(`[CODE-ERROR] - Fallo al consultar los asientos del evento: ${err.message}`);
      throw err;
    }
  }

  /**
   * Reserva una butaca con control estricto de concurrencia (Optimistic Locking)
   */
  public static reserveSeat(seatId: string, userId: string): { reservationId: string; expiresAt: string } {
    try {
      const reservationTransaction = db.transaction(() => {
        // 1. Obtener la butaca actual para verificar version y estado
        const seat = db.prepare(`SELECT id, status, version FROM seats WHERE id = ?`).get(seatId) as any;

        if (!seat) {
          throw new Error(`La butaca especificada '${seatId}' no existe.`);
        }

        if (seat.status !== 'AVAILABLE') {
          throw new ConcurrencyConflictError(`Asiento ya no disponible. Estado actual: '${seat.status}'.`);
        }

        // 2. Intentar actualizar estado de la butaca usando Optimistic Locking (verificando la columna version exacto)
        const updateResult = db.prepare(`
          UPDATE seats 
          SET status = 'RESERVED', version = version + 1 
          WHERE id = ? AND status = 'AVAILABLE' AND version = ?
        `).run(seatId, seat.version);

        if (updateResult.changes === 0) {
          throw new ConcurrencyConflictError('Asiento ya no disponible. Ganado por otra peticion concurrente.');
        }

        // 3. Crear la reserva activa (5 minutos de expiracion)
        const reservationId = `res-${crypto.randomUUID()}`;
        const expiresAtDate = new Date(Date.now() + 5 * 60 * 1000);
        const expiresAtStr = expiresAtDate.toISOString();

        db.prepare(`
          INSERT INTO reservations (id, seat_id, user_id, status, expires_at)
          VALUES (?, ?, ?, 'ACTIVE', ?)
        `).run(reservationId, seatId, userId, expiresAtStr);

        // 4. Registrar auditoria de reserva exitosa
        AuditService.logAction(
          userId,
          'RESERVE_ATTEMPT_SUCCESS',
          `RESERVATION:${reservationId}`,
          `Reserva creada exitosamente para la butaca '${seatId}' con expiracion en ${expiresAtStr}`
        );

        return { reservationId, expiresAt: expiresAtStr };
      });

      return reservationTransaction();
    } catch (err: any) {
      if (err instanceof ConcurrencyConflictError) {
        // Registrar intento fallido en auditoria fuera de la transaccion revertida
        AuditService.logAction(
          userId,
          'RESERVE_ATTEMPT_CONFLICT',
          `SEAT:${seatId}`,
          `Intento de reserva fallido por conflicto de concurrencia: ${err.message}`
        );
        throw err;
      }
      console.error(`[CODE-ERROR] - Error durante el proceso de reserva: ${err.message}`);
      throw err;
    }
  }

  /**
   * Procesa el pago de una reserva en una transaccion ACID atomica
   */
  public static processPayment(reservationId: string, userId: string): { success: boolean; transactionId: string } {
    const paymentTransaction = db.transaction(() => {
      // 1. Obtener y validar la reserva
      const reservation = db.prepare(`
        SELECT r.id, r.seat_id, r.user_id, r.status, r.expires_at, s.version
        FROM reservations r
        JOIN seats s ON r.seat_id = s.id
        WHERE r.id = ?
      `).get(reservationId) as any;

      if (!reservation) {
        throw new Error('Reserva no encontrada.');
      }

      if (reservation.user_id !== userId) {
        throw new Error('La reserva pertenece a otro usuario.');
      }

      if (reservation.status !== 'ACTIVE') {
        throw new Error(`La reserva no se encuentra activa (estado actual: ${reservation.status}).`);
      }

      const now = new Date();
      const expiresAt = new Date(reservation.expires_at);

      if (now > expiresAt) {
        throw new Error('La reserva ha expirado y no puede ser pagada.');
      }

      // 2. Cambiar el estado del asiento a "SOLD" y actualizar la version
      const seatUpdate = db.prepare(`
        UPDATE seats
        SET status = 'SOLD', version = version + 1
        WHERE id = ? AND status = 'RESERVED'
      `).run(reservation.seat_id);

      if (seatUpdate.changes === 0) {
        throw new Error('Fallo al actualizar el estado de la butaca a Vendida.');
      }

      // 3. Marcar reserva como COMPLETED
      const resUpdate = db.prepare(`
        UPDATE reservations
        SET status = 'COMPLETED'
        WHERE id = ? AND status = 'ACTIVE'
      `).run(reservationId);

      if (resUpdate.changes === 0) {
        throw new Error('Fallo al completar la reserva.');
      }

      const transactionId = `tx-${crypto.randomUUID()}`;

      // 4. Asentar auditoria de pago exitoso
      AuditService.logAction(
        userId,
        'PAYMENT_SUCCESS',
        `TRANSACTION:${transactionId}`,
        `Pago simulado procesado exitosamente para la reserva '${reservationId}' en la butaca '${reservation.seat_id}'`
      );

      return { success: true, transactionId };
    });

    try {
      return paymentTransaction();
    } catch (err: any) {
      console.error(`[CODE-ERROR] - Fallo transaccional al procesar el pago: ${err.message}`);
      throw err;
    }
  }

  /**
   * ADMIN: Crea un nuevo evento con sus sectores y asientos numerados en una transaccion ACID
   */
  public static createEventByAdmin(input: CreateEventInput, adminUserId: string): { eventId: string; totalCreatedSeats: number } {
    const createTx = db.transaction(() => {
      const eventId = `evt-${crypto.randomUUID().substring(0, 8)}`;
      
      db.prepare(`
        INSERT INTO events (id, name, description, date, location)
        VALUES (?, ?, ?, ?, ?)
      `).run(eventId, input.name, input.description, input.date, input.location);

      let totalCreatedSeats = 0;

      const insertSector = db.prepare(`
        INSERT INTO sectors (id, event_id, name, price, total_seats)
        VALUES (?, ?, ?, ?, ?)
      `);

      const insertSeat = db.prepare(`
        INSERT INTO seats (id, sector_id, seat_number, status, version)
        VALUES (?, ?, ?, 'AVAILABLE', 1)
      `);

      // Recorrer los sectores enviados por el admin usando bucle idx_tk
      for (let idx_tk = 0; idx_tk < input.sectors.length; idx_tk++) {
        const sec = input.sectors[idx_tk];
        const sectorId = `sec-${crypto.randomUUID().substring(0, 8)}`;
        insertSector.run(sectorId, eventId, sec.name, sec.price, sec.totalSeats);

        // Generar asientos numerados para el sector usando bucle secundario idx_tk_seat
        for (let idx_tk_seat = 1; idx_tk_seat <= sec.totalSeats; idx_tk_seat++) {
          const seatId = `seat-${sectorId}-${idx_tk_seat}`;
          insertSeat.run(seatId, sectorId, idx_tk_seat);
          totalCreatedSeats++;
        }
      }

      // Registrar accion administrativa en Auditoria
      AuditService.logAction(
        adminUserId,
        'ADMIN_EVENT_CREATED',
        `EVENT:${eventId}`,
        `Evento '${input.name}' creado exitosamente con ${input.sectors.length} sectores y ${totalCreatedSeats} asientos.`
      );

      return { eventId, totalCreatedSeats };
    });

    try {
      return createTx();
    } catch (err: any) {
      console.error(`[CODE-ERROR] - Fallo al crear evento administrativo: ${err.message}`);
      throw err;
    }
  }

  /**
   * ADMIN: Obtiene metricas globales del sistema
   */
  public static getAdminStats(): any {
    try {
      const totalEvents = (db.prepare(`SELECT COUNT(*) as c FROM events`).get() as any).c;
      const totalSeats = (db.prepare(`SELECT COUNT(*) as c FROM seats`).get() as any).c;
      const availableSeats = (db.prepare(`SELECT COUNT(*) as c FROM seats WHERE status = 'AVAILABLE'`).get() as any).c;
      const reservedSeats = (db.prepare(`SELECT COUNT(*) as c FROM seats WHERE status = 'RESERVED'`).get() as any).c;
      const soldSeats = (db.prepare(`SELECT COUNT(*) as c FROM seats WHERE status = 'SOLD'`).get() as any).c;

      // Calcular ingresos totales por ventas recaudadas
      const revenueRow = db.prepare(`
        SELECT COALESCE(SUM(sec.price), 0) as totalRevenue
        FROM seats st
        JOIN sectors sec ON st.sector_id = sec.id
        WHERE st.status = 'SOLD'
      `).get() as any;

      const totalAuditLogs = (db.prepare(`SELECT COUNT(*) as c FROM audit_logs`).get() as any).c;

      return {
        totalEvents,
        totalSeats,
        availableSeats,
        reservedSeats,
        soldSeats,
        totalRevenue: revenueRow.totalRevenue,
        totalAuditLogs
      };
    } catch (err: any) {
      console.error(`[CODE-ERROR] - Fallo al consultar estadísticas de administración: ${err.message}`);
      throw err;
    }
  }
}
